raft/raft.go

```diff
@@ -585,24 +594,32 @@ func stepLeader(r *raft, m pb.Message) {
 			r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 	case pb.MsgSnapStatus:
-		if !pr.hasPendingSnapshot() {
+		if pr.State != ProgressStateSnapshot {
 			return
 		}
 		if m.Reject {
 			pr.snapshotFail()
 			raftLogger.Infof("raft: %x snapshot failed, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
+			// QUESTION: change it to stay in snapshot state?
+			// If the network is broken, it will switch between probe and snapshot.
 		} else {
 			pr.snapshotFinish()
 			raftLogger.Infof("raft: %x snapshot succeeded, resumed sending replication messages to %x [%s]", r.id, m.From, pr)
-			// wait for the msgAppResp from the remote node before sending
-			// out the next msgApp
-			pr.waitSet(r.electionTimeout)
 		}
+		pr.becomeProbe()
+		// If snapshot finish, wait for the msgAppResp from the remote node before sending
+		// out the next msgApp.
+		// If snapshot failure, wait for some time before next try
+		pr.waitSet()
 	case pb.MsgUnreachable:
-		if !pr.isUnreachable() {
-			pr.unreachable()
-			raftLogger.Infof("raft: %x failed to send message to %x and changed it to be unreachable [%s]", r.id, m.From, pr)
+		// When in optimistic appending path, if the remote becomes unreachable,
+		// there is big probability that it loses MsgApp. Set next to match + 1
+		// to recover it steadily.
+		if pr.State == ProgressStateReplicate {
```

Comment:
// During optimistic replication, if the remote becomes unreachable,
// there is huge probability that a MsgApp is lost. Set next to match + 1
// to start probing from beginning.